geode-core/src/main/java/org/apache/geode/internal/cache/CacheDistributionAdvisor.java

```diff
@@ -145,14 +153,24 @@ public String toString() {
   private Set<InternalDistributedMember> adviseAllEventsOrCached(final boolean excludeInRecovery)
       throws IllegalStateException {
     getAdvisee().getCancelCriterion().checkCancelInProgress(null);
-    return adviseFilter(profile -> {
-      assert profile instanceof CacheProfile;
-      CacheProfile cp = (CacheProfile) profile;
-      if (excludeInRecovery && cp.inRecovery) {
-        return false;
-      }
-      return cp.cachedOrAllEventsWithListener();
-    });
+
+    // minimize volatile reads by copying ref to local var
+    long tempProfilesVersion = profilesVersion; // volatile read
+    if (adviseAllEventsVersion != tempProfilesVersion || adviseAllEventsVersion == 0
```

Comment:
Instead of having the cached value recomputed based on the state of the boolean excludeInRecovery, would it be better just to support two cached values: one for when excludeInRecovery is false and one for when it is true? I worry that if we have a pattern of asking for both flavors then they will keep throwing away the previous cached value and we end up with no caching.

Reply:
Currently excludeInRecovery is set to true, in both places from where it is called.

Reply:
In that case I think we should remove this parameter since it is an unused feature that is complicating the new implementation.